### PR TITLE
[Feature] Support ReduceLrUpdaterHook

### PR DESCRIPTION
**Pull request by Yshuo-Li**

### CODE REVIEW

File: mmedit/core/scheduler/lr_updater.py [hunk 1]
```diff
@@ -50,3 +50,246 @@ def get_lr(self, runner, base_lr):
...
+                 **kwargs):
+
+        self.val_metric = val_metric
+
+        if mode not in ['min', 'max']:
+            raise ValueError(
+                'mode must be one of "min" or "max", instead got {mode}')
+        self.mode = mode
+
+        if factor >= 1.0:
```
Comment (wangruohui) [inline mmedit/core/scheduler/lr_updater.py:121]: maybe `if factor >= 1.0 or factor <0` is better?

File: mmedit/core/scheduler/lr_updater.py [hunk 2]
```diff
@@ -50,3 +50,246 @@ def get_lr(self, runner, base_lr):
...
+                self.num_bad_epochs = 0
+            else:
+                self.num_bad_epochs += 1
+
+            if self.in_cooldown:
+                self.cooldown_counter -= 1
+                self.num_bad_epochs = 0
+            print(f'train_iter--current {current:.6f} best {self.best:.6f}, '
+                  f'num_bad_epochs {self.num_bad_epochs}, '
+                  f'cooldown {self.in_cooldown} {self.cooldown_counter}')
```
Comment (wangruohui) [inline mmedit/core/scheduler/lr_updater.py:252]: If `val_metric` is specified, we check its value after training iteration based on **training loss**, else, we check it after validation iteration based on some **validation metrics**. Is it right?

We should make this clear in the docstring, otherwise, it's confusing, because it may engage in different hooks.

File: tests/test_runtime/test_lr_updater.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+def test_linear_lr_updater_hook():
+    fake_runner = FakeRunner()
+
+    lr_updater = LinearLrUpdaterHook(by_epoch=False)
+    lr_updater.get_lr(fake_runner, 1)
+
+    lr_updater = LinearLrUpdaterHook(by_epoch=True)
+    lr_updater.get_lr(fake_runner, 1)
+    lr_updater.start = 10
+    lr_updater.get_lr(fake_runner, 1)
```
Comment (wangruohui) [inline tests/test_runtime/test_lr_updater.py:34]: We need to at least assert something in the test, e.g. if the learning rate reduces to a desired value.
Same for `test_reduce_lr_updater_hook`

Comment (wangruohui) [reply tests/test_runtime/test_lr_updater.py:34]: Is this another feature irrelevant to reducing on the plateau?

Comment (Yshuo-Li) [reply tests/test_runtime/test_lr_updater.py:34]: > Is this another feature irrelevant to reducing on the plateau?

Yes, it is another feature irrelevant to reducing on the plateau, which is in lack of unit test

**Comment by wangruohui** [comment]
We still need to assert the `lr` get from `lr_updater.get_lr(fake_runner, 1)` is expected. They are missing currently.

**Review by wangruohui — APPROVED** [review]
